contracts/ExponentialStaking.sol

```diff
@@ -0,0 +1,269 @@
+///
+/// The balance received for staking (and thus the voting power and rewards
+/// distribution) goes up exponentially by the end of the staked period.
+contract ExponentialStaking is ERC20Votes {
+    uint256 public immutable epoch; // timestamp
+    ERC20 public immutable asset; // Must not allow reentrancy
+    RewardsSource public immutable rewardsSource;
+    uint256 public immutable minStakeDuration; // in seconds
+    uint256 public constant maxStakeDuration = 365 days;
+    uint256 constant YEAR_BASE = 14e17;
+    int256 constant NEW_STAKE = -1;
+
+    // 2. Staking and Lockup Storage
+    struct Lockup {
+        uint128 amount;
+        uint128 end;
+        uint256 points;
+    }
+
+    mapping(address => Lockup[]) public lockups;
+
+    // 3. Reward Storage
+    mapping(address => uint256) public rewardDebtPerShare;
+    uint256 public accRewardPerShare;
+
+    // Events
+    event Stake(address indexed user, uint256 lockupId, uint256 amount, uint256 end, uint256 points);
+    event Unstake(address indexed user, uint256 lockupId, uint256 amount, uint256 end, uint256 points);
+    event Reward(address indexed user, uint256 amount);
+    event Penalty(address indexed user, uint256 amount);
+
+    // Core ERC20 Functions
+
+    constructor(address asset_, uint256 epoch_, uint256 minStakeDuration_, address rewardsSource_)
+        ERC20("", "")
+        ERC20Permit("xOGN")
+    {
+        asset = ERC20(asset_);
+        epoch = epoch_;
+        minStakeDuration = minStakeDuration_;
+        rewardsSource = RewardsSource(rewardsSource_);
+    }
+
+    function name() public pure override returns (string memory) {
+        return "Staked OGN";
+    }
+
+    function symbol() public pure override returns (string memory) {
+        return "xOGN";
+    }
+
+    function transfer(address, uint256) public override returns (bool) {
+        revert("Staking: Transfers disabled");
+    }
+
+    function transferFrom(address, address, uint256) public override returns (bool) {
+        revert("Staking: Transfers disabled");
+    }
+
+    // Staking Functions
+
+    /// @notice Stake asset to an address that may not be the same as the
+    /// sender of the funds. This can be used to give staked funds to someone
+    /// else.
+    ///
+    /// If staking before the start of staking (epoch), then the lockup start
+    /// and end dates are shifted forward so that the lockup starts at the
+    /// epoch.
+    ///
+    /// Any rewards previously earned will be paid out or rolled into the stake.
+    ///
+    /// @param amountIn asset to lockup in the stake
+    /// @param duration in seconds for the stake
+    /// @param to address to receive ownership of the stake
+    /// @param stakeRewards should pending user rewards be added to the stake
+    /// @param lockupId previous stake to extend / add funds to. -1 to create a new stake.
+    function stake(uint256 amountIn, uint256 duration, address to, bool stakeRewards, int256 lockupId) external {
```

Comment:
I just realized that we also need an additional parameter to set the delegate (especially when it's gonna be called by the migrator). Let me pull this changeset to my branch and add that (or we could just let the user make two calls)

Reply:
Pushed a minor fix for this: https://github.com/OriginProtocol/ousd-governance/pull/409/commits/3f51c2d70308499b4981396bbae19bb4920a6348

Reply:
Nvm, I didn't see that it needed a typecast. Reverted that and have casted it properly. Also, fixed a few casing issues, it should work fine on CI now